### PR TITLE
Introduce new form object

### PR DESCRIPTION
**Pull request by pyromaniac**
Detaching forms from operations and improving the experience

### CODE REVIEW

**Review by StoneFrog — COMMENTED** [review]
I don't see anything obvious, though I'm not familiar with codebase so that may be the reason. Had few questions though

File: README.md [hunk 1]
```diff
@@ -912,23 +919,79 @@ end
...
+  param :params_map
+
+  def call(_form_class, params, **_context)
+    params.transform_keys { |key| params_map[key] || key }
+  end
+end
+
+# And this will transform nested attributes hash from the form to an array acceptable by the operation:
+# from
+#   `{ id: 42, sections: { '0' => { id: 1, content: "First paragraph" }, 'new' => { content: 'New Paragraph' } } }`
```
[the remark this answers is not in the document]
Comment (pyromaniac) [reply README.md:982]: yes, you are correct, thanks

File: lib/operations/form/base.rb [hunk 2]
```diff
@@ -91,23 +106,43 @@ def assigned_attributes
...
     def model_name
-      ActiveModel::Name.new(self.class)
+      self.class.model_name
     end
 
-    # This should return false if we want to use POST.
-    # Now it is going to generate PATCH form.
     def persisted?
-      true
+      !has_attribute?(self.class.primary_key) || read_attribute(self.class.primary_key).present?
```
[the remark this answers is not in the document]
Comment (pyromaniac) [reply lib/operations/form/base.rb:135]: It is more for compatibility with the existing behavior. `true` is hardcoded in master, so I'm kinda keeping. But any suggestions are welcome.

File: lib/operations/form/base.rb [hunk 3]
```diff
@@ -208,6 +244,19 @@ def wrap_object(path, data, form)
         data
       end
     end
+
+    def build_attribute_name(name)
+      name.to_s.sub(BUILD_ASSOCIATION_PREFIX, "").to_sym if name.to_s.match?(BUILD_ASSOCIATION_PREFIX)
```
[the remark this answers is not in the document]
Comment (pyromaniac) [reply lib/operations/form/base.rb:249]: Should be the same but for `start_with?` a string argument is used and I can't reuse it for sub. Though maybe, use `delete_suffix` :/

But then this is more or less the same. Why do you think regexp is not safe though?

[the remark this answers is not in the document]
Comment (StoneFrog) [reply lib/operations/form/base.rb:249]: ah didn't notice the `\A`. So not that it's not safe, but `name.to_s.start_with?(BUILD_ASSOCIATION_PREFIX)` already tells me exactly what happens. For match I need to scroll all the way up to double check, so jjust a little bit clearer

> Should be the same but for start_with? a string argument is used and I can't reuse it for sub.

It takes regex and string actually.

File: lib/operations/form/base.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (StoneFrog) [reply lib/operations/form/base.rb:135]: initially it was a bit counterintuitive that id => nil is not persisted, but no id means it's persisted. The more I think, the more unsure I am how this should work, so I was hoping we have some real example

[the remark this answers is not in the document]
Comment (pyromaniac) [reply lib/operations/form/base.rb:135]: The real example here is the topmost form object. It usually doesn't have an id, it rather have entity_id. And nested objects are normally having an id. So for nested objects nested attributes forms will be able to figure out their magic and on the top level all we need it for is to determine that the form is going to be POST or PUT.

On the other hand, if we allow only to create nested attributes, there will be no id and maybe this should indeed mean that the object is not persisted. But then we need to find a way to make it compatible with the current behavior.

Or call it a breaking change in the new forms :/ Let me thing of it a bit.

File: spec/operations/form/base_spec.rb [hunk 4]
```diff
@@ -199,11 +250,87 @@
...
       specify { expect(form.model_name).to be_a(ActiveModel::Name) }
       specify { expect(form.model_name.to_s).to eq "Dummy::Form" }
     end
 
+    describe "#persisted?" do
+      subject(:persisted?) { form.persisted? }
+
+      it { is_expected.to be true }
+
+      context "when form has primary_key" do
```
Comment (adamgrad) [inline spec/operations/form/base_spec.rb:303]: I was briefly browsing the setup, but couldn't see it - where is it actually specified in the setup that there is a primary key on this form?

**Review by adamgrad — COMMENTED** [review]
Some interesting changes! Left minor comments in the PR.
I do also have a question regarding your initial message regarding this change:
> Later, all the form-related stuff will be removed from operation definition and forms might even go to a separate gem to keep things modular.

I understand the argument behind modularity, however, what would be other benefits of moving forms outside of operations? Those forms seem like very specific handling for the operations or anything with the same interface like the operations. So what would be the benefit of using those forms without using operations gem?

**Comment by pyromaniac** [comment]
> what would be other benefits of moving forms outside of operations? 

@adamgrad the forms are not only specific to the operations but also to Rails and AR. They support nested attributes and rails strong params etc.

So the idea is that in API-like apps you don't need the forms at all, so why have it? And in the non-rails app with operations the form implementation might be different.

File: spec/operations/form/base_spec.rb (hunk 4 above)
Comment (pyromaniac) [reply spec/operations/form/base_spec.rb:303]: https://github.com/BookingSync/operations/pull/47/files#diff-015981d187060e2473315e3fed1c6480e985502fc8ef8c19c942b137ca0eafbdR25 here

File: lib/operations/form/base.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (pyromaniac) [reply lib/operations/form/base.rb:135]: Addressed in https://github.com/BookingSync/operations/pull/48

**Pull request merged**